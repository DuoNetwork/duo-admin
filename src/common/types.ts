import { ChildProcess } from 'child_process';
export * from '../../../duo-contract-wrapper/src/types';

export interface IBaseMarketData {
	source: string;
	base: string;
	quote: string;
	timestamp: number;
}

export interface ITrade extends IBaseMarketData {
	id: string;
	price: number;
	amount: number;
}

export interface IPriceFix extends IBaseMarketData {
	price: number;
	volume: number;
}

export interface IPrice extends IBaseMarketData {
	period: number;
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
}

export interface IBaseEvent {
	contractAddress: string;
	timestamp: number;
	transactionHash: string;
	blockNumber: number;
}

export interface IAcceptedPrice extends IBaseEvent {
	price: number;
	navA: number;
	navB: number;
}

export interface ITotalSupply extends IBaseEvent {
	tokenA: number;
	tokenB: number;
}

export interface IConversion extends IBaseEvent {
	type: string;
	eth: number;
	tokenA: number;
	tokenB: number;
	fee: number;
	pending?: boolean;
	reverted?: boolean;
}

export interface IStatus {
	process: string;
	timestamp: number;
}

export interface IPriceStatus extends IStatus {
	price: number;
	volume: number;
}

export interface INodeStatus extends IStatus {
	block: number;
}

export interface IAccount {
	address: string;
	privateKey: string;
}

export interface IKey {
	publicKey: string;
	privateKey: string;
}

export interface ISqlAuth {
	host: string;
	user: string;
	password: string;
}

export interface IOption {
	forceREST: boolean;
	pair: string;
	sources: string[];
	exSources: string[];
	assets: string[];
	live: boolean;
	dbLive: boolean;
	server: boolean;
	dynamo: boolean;
	aws: boolean;
	gcp: boolean;
	azure: boolean;
	force: boolean;
	gasPrice: number;
	gasLimit: number;
	source: string;
	event: string;
	provider: string;
	period: number;
	base: string;
	quote: string;
	contractType: string;
	tenor: string;
}

export interface ISourceSettings {
	priceInversed: boolean;
	quoteInversed: boolean;
	separator: string;
	isLowercase: boolean;
	tradesInterval: number;
	supportWS: boolean;
	filterByTimestamp: boolean;
	wsLink: string;
}

export interface ISourceAsset {
	base: {
		[code: string]: boolean;
	};
	quote: {
		[code: string]: boolean;
	};
	mapping: string;
	name: string;
}
export interface ISource {
	assets: {
		[code: string]: ISourceAsset;
	};
	settings: ISourceSettings;
}

export interface IBitfinexRawTradeWS {
	timestamp: number;
	seq: string;
	id: number;
	price: string;
	amount: string;
}

export interface IBitstampRawTradeRest {
	date: number;
	tid: number;
	price: number;
	type: string;
	amount: number;
}

export interface IBitstampRawTradeWs {
	microtimestamp: string;
	amount: number;
	buy_order_id: number;
	sell_order_id: number;
	amount_str: string;
	price_str: string;
	timestamp: string;
	price: number;
	type: number;
	id: number;
}

export interface IBitfinexRawTradeRest {
	timestamp: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeRest {
	timestamp: number;
	timestampms: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeWs {
	type: string;
	eventid: number;
	socket_sequence: number;
	timestamp: number;
	timestampms: number;
	events: IGeminiTradeData[];
}

export interface IGeminiTradeData {
	type: string;
	tid: number;
	price: string;
	amount: string;
	makerSide: string;
}

export type IKrakenRawTrade = IKrakenRawTradeEntry & {
	last: string;
};

export interface IKrakenRawTradeEntry {
	[pair: string]: string[][];
}

export interface IGdaxRawTradeRest {
	error: string[];
	result: {
		[code: string]: Array<Array<string | number>>;
	};
}

export interface ISubProcess {
	source: string;
	instance: ChildProcess;
	lastFailTimestamp: number;
	failCount: number;
}
