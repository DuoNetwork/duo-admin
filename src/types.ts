export interface Trade {
	source: string;
	tradeId: string;
	price: string;
	amount: string;
	tradeType: string;
	sourceTimestamp: string;
}

export interface Price {
	price: string;
	timestamp: string;
}