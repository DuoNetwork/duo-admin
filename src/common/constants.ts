export * from '../../../duo-contract-wrapper/src/constants';
export const TRADES = 'trades';
export const COMMIT = 'commit';
export const SUBSCRIBE = 'subscribe';
export const MINUTELY = 'minutely';
export const HOURLY = 'hourly';
export const NODE = 'node';
export const CLEAN_DB = 'cleanDB';
export const FETCH_PRICE = 'fetchPrice';
export const API_BITFINEX = 'bitfinex';
export const API_GEMINI = 'gemini';
export const API_KRAKEN = 'kraken';
export const API_GDAX = 'gdax';

export const AWS_DYNAMO_API_VERSION = '2012-10-08';
// db setting
export const DB_SQL_SCHEMA_PRICEFEED = 'priceFeedDB';
export const DB_SQL_TRADE = 'trades';
export const DB_SQL_HISTORY = 'historical_price';

export const DB_DUO = 'duo';
export const DB_TRADES = 'trades';
export const DB_LIVE = 'live';
export const DB_DEV = 'dev';
export const DB_PRICES = 'prices';
export const DB_AWS_TRADES_LIVE = 'duo.trades.live';
export const DB_AWS_TRADES_DEV = 'duo.trades.dev';
export const DB_AWS_HOURLY_LIVE = 'duo.prices.60.live';
export const DB_AWS_HOURLY_DEV = 'duo.prices.60.dev';
export const DB_AWS_MINUTELY_LIVE = 'duo.prices.1.live';
export const DB_AWS_MINUTELY_DEV = 'duo.prices.1.dev';
export const DB_AWS_EVENTS_LIVE = 'duo.events.live';
export const DB_AWS_EVENTS_DEV = 'duo.events.dev';
export const DB_AWS_STATUS_LIVE = 'duo.status.live';
export const DB_AWS_STATUS_DEV = 'duo.status.dev';
export const DB_AWS_UI_EVENTS_LIVE = 'duo.uiEvents.live';
export const DB_AWS_UI_EVENTS_DEV = 'duo.uiEvents.dev';
export const DB_TX_QTE = 'quote';
export const DB_TX_BASE = 'base';
export const DB_TX_QUOTE_BASE_ID = 'quoteBaseId';
export const DB_TX_SRC = 'source';
export const DB_TX_ID = 'id';
export const DB_TX_PRICE = 'price';
export const DB_TX_AMOUNT = 'amount';
export const DB_TX_TS = 'timestamp';
export const DB_TX_SYSTIME = 'systime';
export const DB_TX_PAIR = 'pair';
export const DB_UPDATED_AT = 'updatedAt';
export const DB_TX_SRC_DHM = 'sourceDateHourMinute';
export const DB_HISTORY_PRICE = 'price';
export const DB_HISTORY_TIMESTAMP = 'timestamp';
export const DB_QUOTE_BASE_TS = 'quoteBaseTimestamp';
export const DB_HISTORY_VOLUME = 'volume';
export const DB_SRC_DHM = 'sourceDateHourMinute';
export const DB_SRC_DH = 'sourceDateHour';
export const DB_SRC_DATE = 'sourceDate';
export const DB_SRC_YM = 'sourceYearMonth';
export const DB_MN_MINUTE = 'minute';
export const DB_OHLC_OPEN = 'open';
export const DB_OHLC_HIGH = 'high';
export const DB_OHLC_LOW = 'low';
export const DB_OHLC_CLOSE = 'close';
export const DB_OHLC_VOLUME = 'volume';
export const DB_OHLC_TS = 'timestamp';
export const DB_ST_PROCESS = 'process';
export const DB_ST_TS = 'timestamp';
export const DB_ST_BLOCK = 'block';
export const DB_ST_SYSTIME = 'systime';
export const DB_EV_KEY = 'eventKey';
export const DB_EV_TIMESTAMP_ID = 'timestampId';
export const DB_EV_SYSTIME = 'systime';
export const DB_EV_BLOCK_HASH = 'blockHash';
export const DB_EV_BLOCK_NO = 'blockNumber';
export const DB_EV_TX_HASH = 'transactionHash';
export const DB_EV_LOG_STATUS = 'logStatus';
export const DB_EV_PX = 'priceInWei';
export const DB_EV_TS = 'timeInSecond';
export const DB_EV_NAV_A = 'navAInWei';
export const DB_EV_NAV_B = 'navBInWei';
export const DB_EV_TOKEN_A = 'tokenAInWei';
export const DB_EV_TOKEN_B = 'tokenBInWei';
export const DB_EV_ETH = 'ethAmtInWei';
export const DB_EV_ETH_FEE = 'ethFeeInWei';
export const DB_EV_DUO_FEE = 'duoFeeInWei';
export const DB_EV_TOTAL_SUPPLY_A = 'totalSupplyA';
export const DB_EV_TOTAL_SUPPLY_B = 'totalSupplyB';
export const DB_EV_UI_ETH = 'eth';
export const DB_EV_UI_TOKEN_A = 'tokenA';
export const DB_EV_UI_TOKEN_B = 'tokenB';
export const DB_EV_UI_ETH_FEE = 'ethFee';
export const DB_EV_UI_DUO_FEE = 'duoFee';
export const DB_STATUS_EVENT_PUBLIC_OTHERS = 'EVENT_AWS_PUBLIC_OTHERS';

export const STATUS_INTERVAL = 30;
export const TRADES_STATUS_LAST_UPDATE_INTERVAL_WS = 10;

export const AWS_DYNAMO_ROLE_TRADE = 'trade';
export const AWS_DYNAMO_ROLE_HOURLY = 'hourly';
export const AWS_DYNAMO_ROLE_MINUTELY = 'minutely';
export const AWS_DYNAMO_ROLE_EVENT = 'event';
export const AWS_DYNAMO_ROLE_STATUS = 'status';

export const API_FETCH_TRADE_INTERVAL = 1;
export const LOG_INFO = 'INFO';
export const LOG_ERROR = 'ERROR';
export const LOG_DEBUG = 'DEBUG';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const DB_PRICES_PRIMARY_KEY_RESOLUTION: {
	[period: number]: 'minute' | 'hour' | 'day' | 'month';
} = {
	0: 'minute',
	1: 'hour',
	10: 'hour',
	60: 'day',
	360: 'month',
	1440: 'month'
};

export const API_GMN_BASE_URL = 'https://api.gemini.com';
export const API_GMN_WS_LINK = 'wss://api.gemini.com';
export const API_GMN_VERSION = '/v1';
export const API_GMN_ORDER_BOOK = '/book';
export const API_GMN_TRADE = '/trades';
export const API_BFX_BASE_URL = 'https://api.bitfinex.com';
export const API_BFX_VERSION = '/v1';
export const API_BFX_ORDER_BOOK = '/book/';
export const API_BFX_TRADE = '/trades/';
export const API_KRK_BASE_URL = 'https://api.kraken.com';
export const API_KRK_VERSION = '/0/public';
export const API_KRK_ORDER_BOOK = '/Depth';
export const API_KRK_TRADE = '/Trades';
export const API_GDAX_BASE_URL = 'https://api.gdax.com:443/products';
export const API_GDAX_TRADE = '/trades';

export const API_LIST = [API_KRAKEN, API_GEMINI, API_GDAX];
export const EXCHANGE_WEIGHTAGE_TH: { [index: number]: number[] } = {
	4: [0.35, 0.3, 0.25, 0.2],
	3: [0.55, 0.45, 0.35],
	2: [0.65, 0.5]
};
