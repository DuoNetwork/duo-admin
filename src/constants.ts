export const REDRAW_UPDATE_THRESHOLD = 10;

export const EXCHANGE_BITFINEX = 'BITFINEX';
export const EXCHANGE_GEMINI = 'GEMINI';
export const EXCHANGE_KRAKEN = 'KRAKEN';
export const EXCHANGE_GDAX = 'GDAX';

// db setting
export const DB_SQL_SCHEMA_PRICEFEED = 'priceFeedDB';
export const DB_SQL_TRADE = 'eth_trades';
export const DB_SQL_HISTORY = 'eth_historical_price';
export const DB_AWS_TRADES_LIVE = 'trades_live';
export const DB_AWS_TRADES_DEV = 'trades_dev';
export const DB_AWS_HOURLY_LIVE = 'hourly_live';
export const DB_AWS_HOURLY_DEV = 'hourly_dev';
export const DB_AWS_MINUTELY_LIVE = 'minutely_live';
export const DB_AWS_MINUTELY_DEV = 'minutely_dev';
export const DB_AWS_EVENTS_LIVE = 'events_live';
export const DB_AWS_EVENTS_DEV = 'events_dev';
export const DB_AWS_STATUS_LIVE = 'status_live';
export const DB_AWS_STATUS_DEV = 'status_dev';
export const AWS_DYNAMO_API_VERSION = '2012-10-08';

export const AWS_DYNAMO_ROLE_TRADE = 'trade';
export const AWS_DYNAMO_ROLE_HOURLY = 'hourly';
export const AWS_DYNAMO_ROLE_MINUTELY = 'minutely';
export const AWS_DYNAMO_ROLE_EVENT = 'event';
export const AWS_DYNAMO_ROLE_STATUS = 'status';

// wallet and contract
export const DUO_CONTRACT_ADDR = '0x7f43B2Dcb3835F137bAc46DC51847a22893B5D9b';
export const CUSTODIAN_ADDR = '0xe5B47c9B13c31aF3EcbbD06AF743bD1ef91617Ea';
export const A_CONTRACT_ADDR = '0x58f498d914F76327017061147DdDE616C7409214';
export const B_CONTRACT_ADDR = '0xaC12aB42E36FE1F48F514419ca13c90911BD0305';
export const INCEPTION_BLK = 7601741;
export const EVENT_ACCEPT_PRICE = 'AcceptPrice';
export const EVENT_START_PRE_RESET = 'StartPreReset';
export const EVENT_START_RESET = 'StartReset';
export const EVENT_START_TRADING = 'StartTrading';
export const EVENT_CREATE = 'Create';
export const EVENT_REDEEM = 'Redeem';
export const EVENT_COMMIT_PRICE = 'CommitPrice';
export const EVENT_TRANSFER = 'Transfer';
export const EVENT_APPROVAL = 'Approval';
export const EVENT_ADD_ADDRESS = 'AddAddress';
export const EVENT_UPDATE_ADDRESS = 'UpdateAddress';
export const EVENT_REMOVE_ADDRESS = 'RemoveAddress';
export const EVENT_SET_VALUE = 'SetValue';
export const EVENT_COLLECT_FEE = 'CollectFee';
export const OTHER_EVENTS = [
	EVENT_START_TRADING,
	EVENT_ACCEPT_PRICE,
	EVENT_CREATE,
	EVENT_REDEEM,
	EVENT_COMMIT_PRICE,
	EVENT_TRANSFER,
	EVENT_APPROVAL,
	EVENT_ADD_ADDRESS,
	EVENT_UPDATE_ADDRESS,
	EVENT_REMOVE_ADDRESS,
	EVENT_SET_VALUE,
	EVENT_COLLECT_FEE
];

export const STATE_INCEPTION = '0';
export const STATE_TRADING = '1';
export const STATE_PRERESET = '2';
export const STATE_DOWN_RESET = '3';
export const STATE_UP_RESET = '4';
export const STATE_PERIOD_RESET = '5';
export const FN_START_CONTRACT = 'startContract';
export const FN_COMMIT_PRICE = 'commitPrice';
export const SRC_MYETHER = 'myether';
export const SRC_INFURA = 'infura';
export const PROVIDER_LOCAL_HTTP = 'http://localhost:8545';
export const PROVIDER_LOCAL_WS = 'ws://localhost:8546';
export const PROVIDER_MYETHER_LIVE = 'https://api.myetherapi.com/eth';
export const PROVIDER_MYETHER_DEV = 'https://api.myetherapi.com/rop';
export const PROVIDER_INFURA_LIVE = 'https://mainnet.infura.io/WSDscoNUvMiL1M7TvMNP';
export const PROVIDER_INFURA_DEV = 'https://kovan.infura.io/WSDscoNUvMiL1M7TvMNP';
export const TRANSFER_TOKEN_INTERVAL = 5; // in seconds
export const TRANSFER_TOKEN_GAS = 120000;
export const TRANSFER_TOKEN_GAS_TH = 20000000000000000;
export const TRANSFER_INTERVAL = 5; // in seconds
export const TRANSFER_GAS_TH = 20000000000000000;
export const REDEEM_INTERVAL = 5; // in seconds
export const REDEEM_GAS = 80000;
export const REDEEM_GAS_TH = 5000000000000000;
export const CREATE_INTERVAL = 5; // in seconds
export const CREATE_GAS = 200000;
export const CREATE_GAS_TH = 10000000000000000;
export const SYS_STATES = {
	0: 'alphaInBP',
	1: 'betaInWei',
	2: 'feeAccumulatedInWei',
	3: 'periodCouponInWei',
	4: 'limitPeriodicInWei',
	5: 'limitUpperInWei',
	6: 'limitLowerInWei',
	7: 'commissionRateInBP',
	8: 'period',
	9: 'iterationGasThreshold',
	10: 'ethDuoFeeRatio',
	11: 'preResetWaitingBlocks',
	12: 'priceTolInBP',
	13: 'priceFeedTolInBP',
	14: 'priceFeedTimeTol',
	15: 'priceUpdateCoolDown',
	16: 'numOfPrices',
	17: 'nextResetAddrIndex',
	18: 'usersLength',
	19: 'addrPoolLength',
	20: 'lastAdminTime',
	21: 'adminCoolDown'
};

// priceFeed
export const DB_TX_SRC = 'source';
export const DB_TX_ID = 'id';
export const DB_TX_PRICE = 'price';
export const DB_TX_AMOUNT = 'amount';
export const DB_TX_TS = 'timestamp';
export const DB_TX_SYSTIME = 'systime';
export const DB_TX_SRC_DHM = 'sourceDateHourMinute';
export const DB_HISTORY_PRICE = 'price';
export const DB_HISTORY_TIMESTAMP = 'timestamp';
export const DB_HISTORY_VOLUME = 'volume';
export const DB_HR_SRC_DATE = 'sourceDate';
export const DB_HR_HOUR = 'hour';
export const DB_MN_SRC_DATE_HOUR = 'sourceDateHour';
export const DB_MN_MINUTE = 'minute';
export const DB_OHLC_OPEN = 'open';
export const DB_OHLC_HIGH = 'high';
export const DB_OHLC_LOW = 'low';
export const DB_OHLC_CLOSE = 'close';
export const DB_OHLC_VOLUME = 'volume';
export const DB_OHLC_TS = 'timestamp';
export const DB_ST_PROCESS = 'process';
export const DB_ST_TS = 'timestamp';
export const DB_ST_BLOCK = 'block';
export const DB_ST_SYSTIME = 'systime';
export const DB_EVENT_KEY = 'eventKey';
export const DB_EVENT_TIMESTAMP_ID = 'timestampId';
export const DB_EVENT_SYSTIME = 'systime';
export const DB_EVENT_BLOCK_HASH = 'blockHash';
export const DB_EVENT_BLOCK_NO = 'blockNumber';
export const DB_EVENT_TX_HASH = 'transactionHash';
export const DB_EVENT_LOG_STATUS = 'logStatus';
export const DB_STATUS_EVENT_PUBLIC_OTHERS = 'EVENT_AWS_PUBLIC_OTHERS';
export const EXCHANGES = [EXCHANGE_BITFINEX, EXCHANGE_GEMINI, EXCHANGE_KRAKEN, EXCHANGE_GDAX];
export const EXCHANGE_WEIGHTAGE_TH = {
	4: [0.35, 0.3, 0.25, 0.2],
	3: [0.55, 0.45, 0.35],
	2: [0.65, 0.5]
};

export const DEFAULT_GAS_PRICE = 5e9;
export const PRE_RESET_GAS_LIMIT = 120000;
export const RESET_GAS_LIMIT = 7880000;
export const EVENT_FETCH_BLOCK_INTERVAL = 100;
export const EVENT_FETCH_TIME_INVERVAL = 1200000;
