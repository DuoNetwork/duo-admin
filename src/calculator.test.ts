import calculateor from './calculator';
import * as CST from './constants';
import { Trade } from './types';
const trades: Trade[] = require('./samples/ETHUSDtrades.json');
// console.log(trades);

test('test getVolumeMedianPrice', () => {
	CST.EXCHANGES.forEach( exchange => {
		const exchange_trades: Trade[] = trades.filter(item => item.source === exchange);
		let medianPrice: number;
		let volume: number;
		[medianPrice, volume] = calculateor.getVolumeMedianPrice(exchange_trades);
		// console.log([medianPrice, volume]);
		expect(medianPrice).toMatchSnapshot();
		const volumeSum: number = exchange_trades.reduce((a, b) => a + Number(b.amount), 0 );
		expect(volume).toBe(volumeSum / 2);

	});

});

test('test normalizeWeightage', () => {
	const volumeList: number[] = [ 100, 200, 300, 400];
	expect(calculateor.normalizeWeightage(volumeList)).toMatchSnapshot();
});

test('test modifyWeightage', () => {
	let inputWeightage: number[] = [0.5, 0.4, 0.05, 0.05];
	expect(calculateor.modifyWeightage(inputWeightage)).toMatchSnapshot();

	inputWeightage = [0.7, 0.2, 0.1];
	expect(calculateor.modifyWeightage(inputWeightage)).toMatchSnapshot();


	inputWeightage = [0.8, 0.2];
	expect(calculateor.modifyWeightage(inputWeightage)).toMatchSnapshot();
});

test('test getExchangePriceFix', () => {
	CST.EXCHANGES.forEach( exchange => {
		const exchange_trades: Trade[] = trades.filter(item => item.source === exchange);
		const timestamp = exchange_trades.reduce((min, p) => Number(p.sourceTimestamp) < min ? Number(p.sourceTimestamp) : min, Number(exchange_trades[0].sourceTimestamp));
		expect(calculateor.getExchangePriceFix(exchange_trades, timestamp, exchange )).toMatchSnapshot();
	});

});



