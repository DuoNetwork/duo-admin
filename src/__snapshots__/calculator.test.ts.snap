// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test getExchangePriceFix 1`] = `
Array [
  634.7,
  0.2,
]
`;

exports[`test getExchangePriceFix 2`] = `
Array [
  637,
  3.0010585,
]
`;

exports[`test getExchangePriceFix 3`] = `
Array [
  667,
  0.012,
]
`;

exports[`test getExchangePriceFix 4`] = `
Array [
  633,
  4.335,
]
`;

exports[`test getVolumeMedianPrice 1`] = `634.69`;

exports[`test getVolumeMedianPrice 2`] = `638`;

exports[`test getVolumeMedianPrice 3`] = `666.98`;

exports[`test getVolumeMedianPrice 4`] = `633`;

exports[`test modifyWeightage 1`] = `
Array [
  0.35,
  0.3,
  0.17500000000000004,
  0.17500000000000004,
]
`;

exports[`test modifyWeightage 2`] = `
Array [
  0.55,
  0.29999999999999993,
  0.14999999999999997,
]
`;

exports[`test modifyWeightage 3`] = `
Array [
  0.65,
  0.35,
]
`;

exports[`test normalizeWeightage 1`] = `
Array [
  0.1,
  0.2,
  0.3,
  0.4,
]
`;
